sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -4600,3 +4600,155 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3, 4), 5, 5);
+       [1,2,3,4,5]
+      > SELECT _FUNC_(array(5, 3, 2, 1), 2, 4);
+       [5,4,3,2,1]
+  """,
+  group = "array_funcs",
+  since = "3.4.0")
+case class ArrayInsert(srcArrayExpr: Expression, posExpr: Expression, itemExpr: Expression)
+  extends TernaryExpression with ImplicitCastInputTypes with SupportQueryContext
+    with QueryErrorsBase {
+
+  @transient private lazy val elementType: DataType =
+    srcArrayExpr.dataType.asInstanceOf[ArrayType].elementType
+
+  override def dataType: DataType = srcArrayExpr.dataType
+  override def inputTypes: Seq[AbstractDataType] = {
+    (srcArrayExpr.dataType, posExpr.dataType, itemExpr.dataType) match {
+      case (ArrayType(e1, hasNull), e2: IntegralType, e3) if (e2 != LongType) =>
+        TypeCoercion.findTightestCommonType(e1, e3) match {
+          case Some(dt) => Seq(ArrayType(dt, hasNull), IntegerType, dt)
+          case _ => Seq.empty
+        }
+      case (e1, e2, e3) => Seq.empty
+    }
+    Seq.empty
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    (first.dataType, second.dataType, third.dataType) match {
+      case (_: ArrayType, e2, e3) if e2 != IntegerType =>
+        DataTypeMismatch(
+          errorSubClass = "UNEXPECTED_INPUT_TYPE",
+          messageParameters = Map(
+            "paramIndex" -> "2",
+            "requiredType" -> toSQLType(IntegerType),
+            "inputSql" -> toSQLExpr(second),
+            "inputType" -> toSQLType(second.dataType))
+        )
+      case _ => TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override def eval(input: InternalRow): Any = {
+    val arr = first.eval(input)
+    val pos = second.eval(input)
+    val item = third.eval(input)
+
+    if (arr == null || pos == null) {
+      null
+    } else {
+      nullSafeEval(arr, pos, item)
+    }
+  }
+
+  override def nullSafeEval(arr: Any, pos: Any, item: Any): Any = {
+    val baseArr = arr.asInstanceOf[ArrayData]
+    val posInt = pos.asInstanceOf[Int]
+
+    if (baseArr.numElements() + 1  > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+      throw QueryExecutionErrors.concatArraysWithElementsExceedLimitError(baseArr.numElements() + 1)
+    }
+    val validatedPosInt = this.getValidPosIndexOrThrow(posInt, baseArr.numElements())
+
+    if (validatedPosInt < 0 || validatedPosInt > baseArr.numElements()) {
+      null
+    } else {
+      val newArray = new Array[Any](arr.asInstanceOf[ArrayData].numElements() + 1)
+      arr.asInstanceOf[ArrayData].foreach(elementType, (i, v) =>
+        if (i >= validatedPosInt) {
+          newArray(i + 1) = v
+        } else {
+          newArray(i) = v
+        }
+      )
+      if (item != null) {
+        newArray(validatedPosInt) = item
+      }
+      new GenericArrayData(newArray)
+    }
+  }
+
+  private def getValidPosIndexOrThrow(pos: Int, numArrayElements: Int): Int = {
+    val validIdx = if (pos == 0) {
```

Comment:
https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/api/snowflake.snowpark.functions.array_insert.html

For `pos` is 0, Is the behavior ``inserting at position 0 has the same effect as using array_prepend`` easy to implement?

Reply:
Hey- yes, theoretically array_prepend and array_append should probably be implemented using array_insert

Reply:
Oh sorry- you'd like array_insert(pos=0) to equal array_insert(pos=1) right? Happy to implement; alternatively, we could just make 'pos' be zero indexed instead of one indexed. What are your thoughts?

Reply:
I hope `array_insert(pos=0)` equals  `array_prepend`

Reply:
Got it- amended to make this work (though spark arrays being 1 indexed means the example here is more akin to "inserting at position 1 has the same effect as using array_prepend")

Reply:
Ugh, this is a bit of a trap. If I'm testing array_insert() with 0 I walk away with the impression that it is 0 based.
Why would we want this. An error seems to be much more educational.